gigascience/gigadb-website
# Admin styles - Dataset create and update forms

# Pull request for issue: #1505

This is a pull request for the following functionalities:

* Update styles for all admin pages that look similar to:
  * Create page: http://gigadb.gigasciencejournal.com:9170/adminLink/create
  * Update page: http://gigadb.gigasciencejournal.com:9170/adminLink/update/id/83

## How to test?

Navigate to all similar pages and cross check that styles have been updated

Here's the full list:

- http://gigadb.gigasciencejournal.com:9170/adminDatasetAuthor/create
- http://gigadb.gigasciencejournal.com:9170/adminDatasetSample/create
- http://gigadb.gigasciencejournal.com:9170/adminFile/linkFolder
- http://gigadb.gigasciencejournal.com:9170/adminFile/create
- http://gigadb.gigasciencejournal.com:9170/adminDatasetProject/create
- http://gigadb.gigasciencejournal.com:9170/adminLink/create
- http://gigadb.gigasciencejournal.com:9170/adminRelation/create
- http://gigadb.gigasciencejournal.com:9170/datasetFunder/create
- http://gigadb.gigasciencejournal.com:9170/adminManuscript/create
- http://gigadb.gigasciencejournal.com:9170/adminAuthor/create
- http://gigadb.gigasciencejournal.com:9170/adminSample/create
- http://gigadb.gigasciencejournal.com:9170/adminSpecies/create
- http://gigadb.gigasciencejournal.com:9170/adminProject/create
- http://gigadb.gigasciencejournal.com:9170/adminExternalLink/create
- http://gigadb.gigasciencejournal.com:9170/adminLinkPrefix/create
- http://gigadb.gigasciencejournal.com:9170/funder/create
- http://gigadb.gigasciencejournal.com:9170/attribute/create
- http://gigadb.gigasciencejournal.com:9170/report/index
- http://gigadb.gigasciencejournal.com:9170/adminDatasetType/create
- http://gigadb.gigasciencejournal.com:9170/adminFileType/create
- http://gigadb.gigasciencejournal.com:9170/adminFileFormat/create
- http://gigadb.gigasciencejournal.com:9170/news/create
- http://gigadb.gigasciencejournal.com:9170/rssMessage/create
- http://gigadb.gigasciencejournal.com:9170/adminPublisher/create
- http://gigadb.gigasciencejournal.com:9170/datasetLog/create
- http://gigadb.gigasciencejournal.com:9170/user/update/id/8

(the corresponding update pages look identical for the most part, differing mainly in the breadcrumbs --see next section)

## How have functionalities been implemented?

Implementation follows these steps, for the most part (a small number of pages differ a bit):

* use new layout in Controller:
```
$this->layout = 'new_datasetpage';
```
* Wrap both `create` and `update` pages in `<div class="container">`
* Replace title by `TitleBreadcrumb` (replacing XXX by values that fit the page)
```
// create page
	$this->widget('TitleBreadcrumb', [
		'pageTitle' => 'Create XXX',
		'breadcrumbItems' => [
			['label' => 'Admin', 'href' => '/site/admin'],
			['label' => 'Manage', 'href' => 'admin'],
			['isActive' => true, 'label' => 'Create'],
		]
	]);
// update page
	$this->widget('TitleBreadcrumb', [
		'pageTitle' => 'Update XXX ' . $model->id,
		'breadcrumbItems' => [
			['label' => 'Admin', 'href' => '/site/admin'],
			['label' => 'Manage', 'href' => '/XXX/admin'],
			['isActive' => true, 'label' => 'Update'],
		]
	]);
```
* Remove unused variables from pages (old breadcrumbs and menu, which are not used anywhere)
* Update form as follows for better layout and button styling (every XXX is to be replaced):
```
<div class="section form row">
	<div class="col-md-offset-3 col-md-6">
		<?php $form = $this->beginWidget('CActiveForm', array(
			'id' => 'XXX',
			'enableAjaxValidation' => false,
		)); ?>

		<p class="note">Fields with <span class="required">*</span> are required.</p>

		<?php if ($model->hasErrors()) : ?>
			<div class="alert alert-danger">
				<?php echo $form->errorSummary($model); ?>
			</div>
		<?php endif; ?>

// form content

		<div class="pull-right">
			<a href="/XXX/admin" class="btn background-btn-o">Cancel</a>
			<?php echo CHtml::submitButton($model->isNewRecord ? 'Create' : 'Save', array('class' => 'btn background-btn')); ?>
		</div>

		<?php $this->endWidget(); ?>
	</div>
</div>
```
* replace existing fields by components in `/components/controls` since these have built-in styles and accessibility. NOTE: it would be possible to just update the existing html for every field, but that would involve a huge amount of code repetition hence why I created these custom components.

## Any issues with implementation?

* A couple pages have more nuanced issues, too complex to include them within this (already large) PR: #1507, #1509 , these two tickets should be worked on immediately after this one, as they are required for the affected pages to function
* the involved pages are:
  * http://gigadb.gigasciencejournal.com:9170/adminFile/update/id/446
  * http://gigadb.gigasciencejournal.com:9170/adminAuthor/update/id/14

## Any changes to automated tests?

It was necessary to update some acceptance tests:
* Used data-test attrs to make a few tests that relied on CSS classes less brittle
* Commented out a few tests that require work on #1507 to pass

## Any changes to documentation?

None

## Any technical debt repayment?

None

## Any improvements to CI/CD pipeline?

None
